ml_tools/eolearn/ml_tools/sampling.py

```diff
@@ -60,33 +60,19 @@ def sample_by_values(
     :return: A pair of numpy arrays first one containing row indices and second one containing column indices of sampled
         points.
     """
-    rng = rng or np.random.default_rng()
     if image.ndim != 2:
         raise ValueError(f"Given image has shape {image.shape} but sampling operates only on 2D images")
 
-    uniques, counts = np.unique(image, return_counts=True)
-    value_counts = dict(zip(uniques, counts))
+    rng = rng or np.random.default_rng()
+    rows = np.empty([0], dtype=int)
+    columns = np.empty([0], dtype=int)
 
-    all_rows, all_columns = [], []
     for value, n_samples in n_samples_per_value.items():
-        n_points = value_counts.get(value, 0)
-
-        if n_samples > n_points and not replace:
-            raise ValueError(
-                f"For label {value} the requested amount of samples is {n_samples} but only {n_points} points are "
-                "available."
-            )
+        sample_indices = rng.choice(np.nonzero(image == value), size=n_samples, replace=replace, axis=1)
+        rows = np.concatenate((rows, sample_indices[0]))
+        columns = np.concatenate((columns, sample_indices[1]))
```

Comment:
now we're doing N (which is number of values) concatenations (which are `O(n)`), before we did N appends (which are O(1)) and one concatenation. It does simplify the code a bit i guess. Think about it a bit more